### PR TITLE
State handling

### PR DESCRIPTION
**Pull request by dustincoates**
This PR provides support for state handling on intents.

Background:
 - The Alexa Skills Kit has different states, each with a group of intents
 - For example, you can start with no state and triggering the HelpIntent will describe the overall skill
 - Once you enter a further state (e.g. `STARTED_GAME`) the HelpIntent might change to explain that step
 - The Alexa Skills Kit Node.js SDK handles states by emitting an event that corresponds to `${INTENT_NAME}${CURRENT_STATE}` that [has been registered](https://github.com/alexa/alexa-skills-kit-sdk-for-nodejs/blob/62e43964c457feb22eb4e44dc8eb154c6b0ee703/lib/alexa.js#L235)
 - It does that by looking at the `STATE` property on an object of handlers
   - (Currently `STATE`, but also exported as a property on the packaged module

This PR essentially takes a provided state and adds it to the built object that is then registered. It supports multiple objects (corresponding to multiple states) via a passing of an array of objects *or* a single state via passing a single object.

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -36,18 +37,18 @@ export default function algoliaAlexaAdapter (opts) {
...
-  skill.handler = function(event, context) {
-    const alexa = Alexa.handler(event, context);
+  skill.handler = function(event, context, callback) {
+    const alexa = Alexa.handler(event, context, callback);
     alexa.appId = alexaAppId;
     if (languageStrings) {
       alexa.resources = languageStrings;
     }
-    alexa.registerHandlers(buildHandlers(handlers, index));
+    alexa.registerHandlers(...buildHandlers(handlers, index, _StateString));
```
Comment (bobylito) [inline src/index.js]: Why do we spread the array now?

File: src/utils/build_handlers.js [hunk 2]
```diff
@@ -24,12 +29,36 @@ export default function buildHandlers (handlersObj, index) {
...
 
-      return handlersObj[key];
+      return object;
     } else {
       throw new Error('Intent handler must either be a function or an object ' +
       'with key of "answerWith" which is a function.');
     }
+  }, {});
+
+  delete result.STATE_STRING;
```
Comment (bobylito) [inline src/utils/build_handlers.js]: Something is confusing here with `STATE_STRING`. It is a variable but also here you delete the key that is named `STATE_STRING` right?

File: src/utils/build_handlers.js [hunk 3]
```diff
@@ -1,19 +1,24 @@
...
 }
 
-export default function buildHandlers (handlersObj, index) {
-  Object.keys(handlersObj).map(key => {
-    if (isFunction(handlersObj[key])) {
-      return handlersObj[key];
-    } else if (hasAnswerWith(handlersObj[key])) {
-      const func = handlersObj[key].answerWith;
-      const paramsObj = handlersObj[key].params;
+function buildFromObject(obj, index, STATE_STRING) {
```
Comment (dustincoates) [inline src/utils/build_handlers.js]: Good point--changing

File: src/utils/build_handlers.js (hunk 2 above)
Comment (dustincoates) [inline src/utils/build_handlers.js]: Yes, thanks for catching!

File: src/index.js (hunk 1 above)
Comment (dustincoates) [inline src/index.js]: The ultimate argument should be 1..n objects.

File: src/utils/is_of_type.js [hunk 4]
```diff
@@ -6,5 +6,6 @@ function isOfType(signature) {
 
 const isObject = isOfType('[object Object]');
 const isFunction = isOfType('[object Function]');
+const isArray = isOfType('[object Array]');
```
Comment (dustincoates) [inline src/utils/is_of_type.js]: Indeed.

File: src/utils/build_handlers.js [hunk 5]
```diff
@@ -24,12 +29,34 @@ export default function buildHandlers (handlersObj, index) {
...
+
+function buildFromArray(arr, index, stateString) {
+  return arr.map(obj => buildFromObject(obj, index, stateString)[0]);
+}
+
+export default function buildHandlers (handlers, index, stateString) {
+  let result;
+
+  if (isArray(handlers)) {
+    result = buildFromArray(handlers, index, stateString);
```
Comment (bobylito) [inline src/utils/build_handlers.js]: I'd rather `return` here directly.

**Pull request merged**